modules/src/applications/transfer/coin.rs

```diff
@@ -0,0 +1,157 @@
+use core::fmt::{Display, Error as FmtError, Formatter};
+use core::str::{from_utf8, FromStr};
+use ibc_proto::cosmos::base::v1beta1::Coin as ProtoCoin;
+use safe_regex::regex;
+use serde::{Deserialize, Serialize};
+
+use super::amount::Amount;
+use super::denom::{BaseDenom, PrefixedDenom};
+use super::error::Error;
+use crate::prelude::*;
+use crate::serializers::serde_string;
+
+/// A `Coin` type with fully qualified `PrefixedDenom`.
+pub type PrefixedCoin = Coin<PrefixedDenom>;
+
+/// A `Coin` type with an unprefixed denomination.
+pub type BaseCoin = Coin<BaseDenom>;
+
+pub type RawCoin = Coin<String>;
+
+/// Coin defines a token with a denomination and an amount.
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Coin<D> {
+    /// Denomination
+    pub denom: D,
+    /// Amount
+    #[serde(with = "serde_string")]
+    pub amount: Amount,
+}
+
+impl<D: FromStr> Coin<D>
+where
+    D::Err: Into<Error>,
+{
+    pub fn from_string_list(coin_str: &str) -> Result<Vec<Self>, Error> {
+        coin_str.split(',').map(FromStr::from_str).collect()
+    }
+}
+
+impl<D: FromStr> FromStr for Coin<D>
```

Comment:
I believe this is Cosmos SDK-specific code and is not being used in the modules. Can we move it (and any related tests) out of the modules crate?

Reply:
This is needed for parsing the `IncentivizedPacket` event from `AbciEvent` in #1981. We can remove this after splitting the IBC module away from the relayer code and remove the parsing for ABCI events.

Reply:
Sounds good. :+1: